File: easy/Chardonnay-Or-Cabernet/Chardonnay Or Cabernet.cs
using System;
using System.IO;
using System.Collections.Generic;

class Program
{
    static void Main(string[] args)
    {
        using (StreamReader reader = File.OpenText(args[0]))
        while (!reader.EndOfStream)
        {
            string line = reader.ReadLine();
            if (null == line)
                continue;
            ShowWine(line);
        }
    }
    
    static void ShowWine(string line){
        int pos = line.IndexOf('|');
        string phrase = line.Substring(pos+1).Trim();
        string[] wines = line.Substring(0,pos).Split(' ');
        string result = "";
        foreach(string wine in wines){
            int i = 0;
            string wineTemp = wine;
            while(i<phrase.Length){
                int tempos = wineTemp.IndexOf(phrase[i]);
                if (tempos>=0) wineTemp = wineTemp.Remove(tempos, 1);       
                else break;
                i++;
            }
            if(wine.Length - phrase.Length == wineTemp.Length)result += wine + " ";
        }
        result = result.Trim();
        if(result.Length>0)Console.WriteLine(result);
        else Console.WriteLine("False");
    }
}

File: easy/Chardonnay-Or-Cabernet/Chardonnay Or Cabernet.java
import java.io.*;
public class Main {
    public static void main (String[] args) throws IOException {
        File file = new File(args[0]);
        BufferedReader buffer = new BufferedReader(new FileReader(file));
        String line;
        while ((line = buffer.readLine()) != null) {
            line = line.trim();
            ShowWine(line);
        }
    }
    
    static void ShowWine(String line){
        int pos = line.indexOf("|");
        String phrase = line.substring(pos+1).trim();
        String[] wines = line.substring(0,pos).split(" ");
        String result = "";
        for(String wine : wines){
            int i = 0;
            String wineTemp = wine;
            while(i<phrase.length()){
                int tempos = wineTemp.indexOf(phrase.charAt(i));
                if (tempos>=0) wineTemp = wineTemp.substring(0,tempos) + wineTemp.substring(tempos+1);
                else break;
                i++;
            }
            if(wine.length() - phrase.length() == wineTemp.length()) result += wine + " ";
        }
        result = result.trim();
        System.out.println( result.length()>0 ? result : "False");
    }
}
